arangod/Metrics/ClusterMetricsFeature.cpp

```diff
@@ -0,0 +1,160 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2022 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Valery Mironov
+////////////////////////////////////////////////////////////////////////////////
+
+#include "Metrics/ClusterMetricsFeature.h"
+#include "Cluster/ClusterFeature.h"
+#include "Cluster/ClusterMethods.h"
+#include "Metrics/MetricsFeature.h"
+#include "Network/NetworkFeature.h"
+#include "Scheduler/SchedulerFeature.h"
+
+namespace arangodb::metrics {
+namespace {
+
+using namespace std::chrono_literals;
+constexpr std::chrono::steady_clock::duration kTimeout = 15s;
+
+}  // namespace
+
+ClusterMetricsFeature::ClusterMetricsFeature(
+    application_features::ApplicationServer& server)
+    : ApplicationFeature{server, "ClusterMetrics"} {
+  setOptional();
+  startsAfter<ClusterFeature>();
+  startsAfter<NetworkFeature>();
+  startsAfter<SchedulerFeature>();
+}
+
+void ClusterMetricsFeature::validateOptions(
+    std::shared_ptr<options::ProgramOptions> /*options*/) {
+  if (!ServerState::instance()->isCoordinator()) {
+    disable();
+  }
+}
+
+void ClusterMetricsFeature::asyncUpdate() {
+  if (isEnabled() && _count.fetch_add(1, std::memory_order_acq_rel) == 0) {
+    forceAsyncUpdate();
+  }
+}
+
+void ClusterMetricsFeature::forceAsyncUpdate() {
+  auto internal = [this] {
+    _count.store(1, std::memory_order_relaxed);
+    // TODO try get from agency
+    metricsOnCoordinator(server().getFeature<NetworkFeature>(),
```

Comment:
`metricsOnCoordinator` is not noexcept.  We may miss fetch_sub call and no other `asyncUpdate`  call will schedule new metrics update.  We could use some "scopeGuard" to ensure proper restart in the worst case.

Reply:
Done